src/aws_credentials_ec2.erl

```diff
@@ -26,51 +26,100 @@
 -export([fetch/1]).
 
 -spec fetch(aws_credentials_provider:options()) ->
-        {ok, aws_credentials_provider:credentials(), aws_credentials_provider:expiration()} |
-        {error, any()}.
+        {error, _}
+      | {ok, aws_credentials_provider:credentials(), aws_credentials_provider:expiration()}.
 fetch(_Options) ->
     % Fetch the IMDS session token and convert it to the request headers.
     % Then pass the request headers to subsequent IMDS requests.
-    {ok, SessionToken} = fetch_session_token(),
-    {ok, RequestHeaders} = request_headers(SessionToken),
-    {ok, Role} = fetch_role(RequestHeaders),
-    {ok, AccessKeyID, SecretAccessKey, ExpirationTime, Token} =
-      fetch_metadata(Role, RequestHeaders),
-    {ok, Region} = fetch_document(RequestHeaders),
-    Credentials = aws_credentials:make_map(?MODULE, AccessKeyID, SecretAccessKey, Token, Region),
-    {ok, Credentials, ExpirationTime}.
+    SessionTokenR = fetch_session_token(),
+    RequestHeadersR = request_headers(SessionTokenR),
+    RoleR = fetch_role(RequestHeadersR),
+    MetadataR = fetch_metadata(RoleR, RequestHeadersR),
+    DocumentR = fetch_document(RequestHeadersR),
+    make_map(MetadataR, DocumentR).
 
--spec fetch_session_token() -> {ok, session_token()}.
+-spec fetch_session_token() ->
+        {error, ec2_session_token_unavailable}
+      | {ok, session_token()}.
 fetch_session_token() ->
   RequestHeaders = [{?SESSION_TOKEN_TTL_HEADER, ?SESSION_TOKEN_TTL_SECONDS}],
-  {ok, 200, Body, _Headers} =
-    aws_credentials_httpc:request(put, ?SESSION_TOKEN_URL, RequestHeaders),
-  {ok, Body}.
+  case aws_credentials_httpc:request(put, ?SESSION_TOKEN_URL, RequestHeaders) of
+    {ok, 200, Body, _Headers} -> {ok, Body};
+    _Other -> {error, ec2_session_token_unavailable}
```

Comment:
Would it be helpful to either be stricter with what we accept here or to include `_Other` in the return value?
If this or any other HTTP call here doesn't return 200 it will be hard for the user to know what happened.

Reply:
Got it. I think I can add an exception for that. I'll do it in all paths concerned. Thanks.

Reply:
b679dc3